go/vt/mysqlctl/binlogs_gtid.go

```diff
@@ -0,0 +1,198 @@
+func FindPITRPath(restoreToGTIDSet mysql.GTIDSet, manifests [](*BackupManifest)) (shortestPath [](*BackupManifest), err error) {
+	sortedManifests := make([](*BackupManifest), 0, len(manifests))
+	for _, m := range manifests {
+		if m != nil {
+			sortedManifests = append(sortedManifests, m)
+		}
+	}
+	sort.SliceStable(sortedManifests, func(i, j int) bool {
+		return sortedManifests[j].Position.GTIDSet.Union(sortedManifests[i].PurgedPosition.GTIDSet).Contains(sortedManifests[i].Position.GTIDSet)
+	})
+	mostRelevantFullBackupIndex := -1 // an invalid value
+	for i, manifest := range sortedManifests {
+		if manifest.Incremental {
+			continue
+		}
+		if restoreToGTIDSet.Contains(manifest.Position.GTIDSet) {
+			// This backup is <= desired restore point, therefore it's valid
+			mostRelevantFullBackupIndex = i
+		}
+	}
+
+	if mostRelevantFullBackupIndex < 0 {
+		// No full backup prior to desired restore point...
+		return nil, vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no full backup found before GTID %v", restoreToGTIDSet)
+	}
+	// All that interests us starts with mostRelevantFullBackupIndex: that's where the full backup is,
+	// and any relevant incremental backups follow that point (because manifests are sorted by backup pos, ascending)
+	sortedManifests = sortedManifests[mostRelevantFullBackupIndex:]
+	// Of all relevant backups, we take the most recent one.
+	fullBackup := sortedManifests[0]
+	if restoreToGTIDSet.Equal(fullBackup.Position.GTIDSet) {
+		// Perfect match, we don't need to look for incremental backups.
+		// We just skip the complexity of the followup section.
+		// The result path is a single full backup.
+		return append(shortestPath, fullBackup), nil
+	}
+	purgedGTIDSet := fullBackup.PurgedPosition.GTIDSet
+
+	var validRestorePaths []BackupManifestPath
+	// recursive function that searches for all possible paths:
+	var findPaths func(baseGTIDSet mysql.GTIDSet, pathManifests []*BackupManifest, remainingManifests []*BackupManifest)
+	findPaths = func(baseGTIDSet mysql.GTIDSet, pathManifests []*BackupManifest, remainingManifests []*BackupManifest) {
+		if baseGTIDSet.Contains(restoreToGTIDSet) {
+			// successful end of path. Update list of successful paths
+			validRestorePaths = append(validRestorePaths, pathManifests)
+			return
+		}
+		if len(remainingManifests) == 0 {
+			// end of the road. No possibilities from here.
+			return
+		}
+		// if the next manifest is eligible to be part of the path, try it out
+		if IsValidIncrementalBakcup(baseGTIDSet, purgedGTIDSet, remainingManifests[0]) {
+			nextGTIDSet := baseGTIDSet.Union(remainingManifests[0].Position.GTIDSet)
+			findPaths(nextGTIDSet, append(pathManifests, remainingManifests[0]), remainingManifests[1:])
+		}
+		// also, try without the next manifest
+		findPaths(baseGTIDSet, pathManifests, remainingManifests[1:])
+	}
+	// find all paths, entry point
+	findPaths(fullBackup.Position.GTIDSet, sortedManifests[0:1], sortedManifests[1:])
```

Comment:
Another thing that I noticed from my understanding which might not be a cause for concern now is that this algorithm has a worse-case exponential run-time. For example, if we have a base backup for position [1,50] and we have 10 incremental backups like - [10, 100], [11, 101], [12, 102], [13, 103].... [19, 109] and the users asks us to restore until the position `109`, then all the subsets of the incremental backups containing the start and the final backup are actually valid.
So, the algorithm will end up populating the validRestorePaths array to 2^9 possible paths, only one of which will have the shortest length.

In the future, if we decide to improve the performance specifically for these corner cases, we can use a dynamic programming-based approach. Please disregard this if my understanding of the function is flawed.

Reply:
> then all the subsets of the incremental backups containing the start and the last incremental backup are actually valid.

You're right. In practice I don't see the situation where this happens because typically you will create an incremental backup from the end of the last known good backup. And so typically we expect a linear sequence of backups, or may be two such sequences if backups are taken from two different machines.

Of course reality can always surprise. A dynamic approach sounds good. We can definitely iterate.

Reply:
Gotcha! 👍

Reply:
@GuptaManan100 let me look into it. If things turn to be too complex, then we can always choose the first valid solution, even if suboptimal.

Reply:
Thank-you! 💕

Reply:
@GuptaManan100 in https://github.com/vitessio/vitess/pull/11097/commits/2ea863cd8def8857a06da66f87bdf50dead09e16 we stop the search immediately after finding the first valid restore path, even if it's not the optimal. I've left the rest of the algorithm intact, including the followup `sort` of the restore paths (now there's just one) in case we even want to re-examine.

Tests were adapted (previously validated that shortest path is chosen, now they validate that first path is chosen)

Reply:
Awesome! Thank-you! 👍